servers/zts/src/main/java/com/yahoo/athenz/zts/store/DataStore.java

```diff
@@ -1550,6 +1571,23 @@ public void setCloudStore(CloudStore cloudStore) {
         return domainData.getRoles();
     }
 
+    public List<String> getRolesRequireRoleCert(String principal) {
```

Comment:
This implementation is not going to work at large scale. Image if we have thousands of roles setup with the option. It means for every request we're going to iterate through the list of roles just to find a match. Instead we should have a cache with memberName -> roleName. 

However, maintaining such a cache has its own issues where we need to keep updating the cache when the role is updated so we should revisit this logic and see how we should handle it. Maybe write a small design doc and we can discuss the implementation there.

We must also handle the case where the principal includes wildcards. So if I have a role member as athenz.* then any service within athenz domain should match.

Reply:
Sure thing, working on it.

Reply:
Please see the updated version. I'm keeping cache in 3 data structures:
1. For regular members - same as for groups using the existing methods
2. For wildcard members - Set of all roles where a wildcard is a member
3. For prefix members - Use a prefix trie so finding a match to a given principal is efficient.